tests/roles/run_tests/tasks/upgrade.yml

```diff
@@ -201,7 +201,7 @@
       dest: "{{ overrides }}/infrastructure-metal3/{{ CAPM3_REL_TO_VERSION }}/metadata.yaml"
 
   - name: Perform upgrade on the target cluster
-    ansible.builtin.command: clusterctl upgrade apply --contract v1beta1
+    ansible.builtin.command: clusterctl upgrade apply --core capi-system/cluster-api:v1.2.1 --bootstrap capi-kubeadm-bootstrap-system/kubeadm:v1.2.1 --control-plane capi-kubeadm-control-plane-system/kubeadm:v1.2.1 --infrastructure capm3-system/metal3:v1.2.0-rc.1
```

Comment:
Clarification: to be able to use clusterctl upgrade apply as per CAPI book, specific version has to be provided in the command, xref: https://cluster-api.sigs.k8s.io/clusterctl/commands/upgrade.html#upgrading-to-pre-release-provider-versions